services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -957,6 +1071,169 @@ class HakemusServiceITest(
+            auditLogRepository.deleteAll()
+
+            hakemusService.cancelAndDelete(hakemus, USERNAME)
+
+            assertThat(auditLogRepository.findAll()).single().isSuccess(Operation.DELETE) {
+                hasUserActor(USERNAME, TestUtils.mockedIp)
+                withTarget {
+                    prop(AuditLogTarget::id).isEqualTo(hakemus.id.toString())
+                    prop(AuditLogTarget::type).isEqualTo(ObjectType.HAKEMUS)
+                    hasObjectBefore(hakemus)
+                    prop(AuditLogTarget::objectAfter).isNull()
+                }
+            }
+        }
+
+        @Test
+        fun `cancels the application in Allu before deleting when the application is pending in Allu`() {
+            val hakemus =
+                hakemusFactory.builder().withStatus(ApplicationStatus.PENDING, alluId).save()
+            every { alluClient.getApplicationInformation(alluId) } returns
+                AlluFactory.createAlluApplicationResponse(alluId, ApplicationStatus.PENDING)
+            justRun { alluClient.cancel(alluId) }
+            every { alluClient.sendSystemComment(alluId, any()) } returns 1324
+
+            hakemusService.cancelAndDelete(hakemus, USERNAME)
+
+            assertThat(applicationRepository.findAll()).isEmpty()
+            verifySequence {
+                alluClient.getApplicationInformation(alluId)
+                alluClient.cancel(alluId)
+                alluClient.sendSystemComment(alluId, ALLU_USER_CANCELLATION_MSG)
+            }
+        }
+
+        @Test
+        fun `throws an exception when the application is past pending in Allu`() {
+            val hakemus =
+                hakemusFactory.builder().withStatus(ApplicationStatus.PENDING, alluId).save()
+            every { alluClient.getApplicationInformation(alluId) } returns
+                AlluFactory.createAlluApplicationResponse(alluId, ApplicationStatus.APPROVED)
+
+            val failure = assertFailure { hakemusService.cancelAndDelete(hakemus, USERNAME) }
+
+            failure.all {
+                hasClass(ApplicationAlreadyProcessingException::class)
+                messageContains("id=${hakemus.id}")
+                messageContains("alluId=$alluId")
+            }
+            assertThat(applicationRepository.findAll()).hasSize(1)
+            verifySequence { alluClient.getApplicationInformation(alluId) }
+        }
+
+        @Test
+        fun `deletes yhteystiedot and yhteyshenkilot`() {
+            val hakemus =
+                hakemusFactory
+                    .builder()
+                    .withNoAlluFields()
+                    .hakija()
+                    .tyonSuorittaja()
+                    .rakennuttaja()
+                    .asianhoitaja()
+                    .save()
+
+            hakemusService.cancelAndDelete(hakemus, USERNAME)
+
+            assertThat(applicationRepository.findAll()).isEmpty()
+            assertThat(hakemusyhteystietoRepository.findAll()).isEmpty()
```

Comment:
Should this also check that hankekayttajas are not deleted? And that hakemusyhteyshenkilo are deleted.

Reply:
Yes, it would be useful to test for those. Thank you for the suggestion. Checks added.